clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -173,13 +192,83 @@ void discoverD2Service() {
       if (isServiceDiscovered) {
         return;
       }
-      transportClient.setServiceName(clusterDiscoveryD2ServiceName);
-      String serverD2ServiceName = d2ServiceDiscovery.find(transportClient, storeName, true).getServerD2Service();
-      transportClient.setServiceName(serverD2ServiceName);
+      d2TransportClient.setServiceName(clusterDiscoveryD2ServiceName);
+      String serverD2ServiceName = d2ServiceDiscovery.find(d2TransportClient, storeName, true).getServerD2Service();
+      d2TransportClient.setServiceName(serverD2ServiceName);
       isServiceDiscovered = true;
     }
   }
 
+  /**
+   * Warmup the connection from the client to Instances discovered from the metadata update.
+   * To avoid duplicate warmup requests being sent which might delay the warmup process,
+   * {@link #warmedUpInstances} is used to track the list of already warmed up instances.
+   */
+  private boolean warmupConnectionToInstances(int currentFetchedVersion, int partitionCount)
+      throws ExecutionException, InterruptedException {
+    List<CompletableFuture<?>> futureList = new ArrayList<>();
+    Set<String> allReplicasInCurrentFetchedVersion = new HashSet<>();
+    for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+      allReplicasInCurrentFetchedVersion.addAll(getReplicas(currentFetchedVersion, partitionId));
+    }
+    LOGGER.info(
+        "New fetched metadata has {} instances for store {} version {}",
+        allReplicasInCurrentFetchedVersion.size(),
+        storeName,
+        currentFetchedVersion);
+
+    // Remove all instances in warmedUpInstances that are not in allReplicasInCurrentFetchedVersion
+    // to keep the warmedUpInstances relatively fresh
+    warmedUpInstances.retainAll(allReplicasInCurrentFetchedVersion);
+
+    // Find all the instances in allReplicasInCurrentFetchedVersion that are not in
+    // warmedUpInstances to start warmup for those instances
+    Set<String> newInstances = new HashSet<>(allReplicasInCurrentFetchedVersion);
+    newInstances.removeAll(warmedUpInstances);
+    int instanceNum = newInstances.size();
+    LOGGER.info(
+        "Starting H2 connection warm up from fast client to {} instances for store {} version {}",
+        instanceNum,
+        storeName,
+        currentFetchedVersion);
+    if (instanceNum == 0) {
+      return Boolean.TRUE;
```

Comment:
nit, the function returns the primitive type `boolean`, `return true;` seems natural to avoid the type conversion `Boolean` -> `boolean`, but maybe you have other preference and I think that's fine as well.